### DISCUSSION TITLE
Incorporating Unique Control Vectors for Each TaggedWeightedGaussianState in GM-PHD Filter

### DISCUSSION THREAD

**Discussion by AliDehghani88** [dstl/Stone-Soup#893] [Q&A]
Hello,

I am currently working through the [Gaussian mixture PHD tutorial](https://stonesoup.readthedocs.io/en/latest/auto_tutorials/filters/GMPHDTutorial.html) and have encountered a challenge that I would like your assistance with.

In the [tracker loop](https://stonesoup.readthedocs.io/en/latest/auto_tutorials/filters/GMPHDTutorial.html#run-the-tracker), we have ```current_state = reduced_states```, which contains multiple instances of TaggedWeightedGaussianState, each with its state vector, covariance, weight, tag, and timestamp.

During the hypothesis generation step:

```python
hypotheses = hypothesiser.hypothesise(
    current_state,
    measurements,
    timestamp=time,
    order_by_detection=True
)
```
the Kalman prediction typically uses the formula ```F*X + B*u```. In my case, it's essential to apply distinct control inputs (u) for each TaggedWeightedGaussianState.
```python
X1_k = F * X1_k-1 + B * u1
X2_k = F * X2_k-1 + B * u2
```
I understand that I can set a uniform control vector as follows example:
```python
hypothesiser.hypothesiser.predictor.control_model.control_vector = np.array([[1], [1.1]])
```
However, I'm struggling with a scenario where each TaggedWeightedGaussianState in current_state requires a unique control vector. 

Could you please guide me on how to assign them? 

Thank you

**Comment by jmbarr** [comment]
I've been having similar thoughts on the shortcomings of the ControlModels which, to be honest, weren't put together with a particular use in mind. I think it's better if you use _u_ as an input to control_model.function() to return the control output. That currently works for `LinearControlModel` at least but the trouble is nearly all of the predictors will have to be reengineered to use `.function(u)` instead of `.control_vector`.
